// @parity/hardhat-polkadot/dist/cli/prompt.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.confirmRecommendedDepsInstallation = confirmRecommendedDepsInstallation;
exports.confirmProjectCreation = confirmProjectCreation;
exports.confirmTelemetryConsent = confirmTelemetryConsent;
exports.confirmHHVSCodeInstallation = confirmHHVSCodeInstallation;
function createConfirmationPrompt(name, message) {
    return {
        type: "confirm",
        name,
        message,
        initial: "y",
        default: "(Y/n)",
        isTrue(input) {
            if (typeof input === "string") {
                return input.toLowerCase() === "y";
            }
            return input;
        },
        isFalse(input) {
            if (typeof input === "string") {
                return input.toLowerCase() === "n";
            }
            return input;
        },
        format() {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const that = this;
            const value = that.value === true ? "y" : "n";
            if (that.state.submitted === true) {
                return that.styles.submitted(value);
            }
            return value;
        },
    };
}
async function confirmRecommendedDepsInstallation(depsToInstall, packageManager) {
    const { default: enquirer } = await Promise.resolve().then(() => __importStar(require("enquirer")));
    let responses;
    try {
        responses = await enquirer.prompt([
            createConfirmationPrompt("shouldInstallPlugin", `Do you want to install this sample project's dependencies with ${packageManager} (${Object.keys(depsToInstall).join(" ")})?`),
        ]);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
    }
    catch (e) {
        if (e === "") {
            return false;
        }
        throw e;
    }
    return responses.shouldInstallPlugin;
}
async function confirmProjectCreation() {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const enquirer = require("enquirer");
    return enquirer.prompt([
        {
            name: "projectRoot",
            type: "input",
            initial: process.cwd(),
            message: "Hardhat project root:",
        },
        createConfirmationPrompt("shouldAddGitIgnore", "Do you want to add a .gitignore?"),
    ]);
}
async function confirmTelemetryConsent() {
    return confirmationPromptWithTimeout("telemetryConsent", "Help us improve Hardhat with anonymous crash reports & basic usage data?");
}
async function confirmHHVSCodeInstallation() {
    return confirmationPromptWithTimeout("shouldInstallExtension", "Would you like to install the Hardhat for Visual Studio Code extension? It adds advanced editing assistance for Solidity to VSCode");
}
async function confirmationPromptWithTimeout(name, message, timeoutMilliseconds = 10_000) {
    try {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const enquirer = require("enquirer");
        const prompt = new enquirer.prompts.Confirm(createConfirmationPrompt(name, message));
        let timeout;
        const timeoutPromise = new Promise((resolve) => {
            timeout = setTimeout(resolve, timeoutMilliseconds);
        });
        const result = await Promise.race([prompt.run(), timeoutPromise]);
        clearTimeout(timeout);
        if (result === undefined) {
            await prompt.cancel();
        }
        return result;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
    }
    catch (e) {
        if (e === "") {
            return undefined;
        }
        throw e;
    }
}
