// @parity/hardhat-polkadot-node/dist/core/global-interceptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.interceptAndWrapTasksWithNode = interceptAndWrapTasksWithNode;
const plugins_1 = require("hardhat/plugins");
const utils_1 = require("../utils");
function interceptAndWrapTasksWithNode() {
    const polkadotGlobal = global;
    const taskMap = polkadotGlobal.__hardhatContext.tasksDSL.getTaskDefinitions();
    if (!polkadotGlobal._polkadotTasksForWrapping) {
        return;
    }
    polkadotGlobal._polkadotTasksForWrapping.taskNames.forEach((taskName) => {
        const foundTask = taskMap[taskName];
        if (!foundTask) {
            return;
        }
        if (foundTask.isSubtask) {
            polkadotGlobal.__hardhatContext.tasksDSL.subtask(foundTask.name, foundTask.description, wrapTaskWithNode);
        }
        polkadotGlobal.__hardhatContext.tasksDSL.task(foundTask.name, foundTask.description, wrapTaskWithNode);
    });
}
async function wrapTaskWithNode(taskArgs, env, runSuper) {
    if (env.network.polkavm !== true || env.network.name !== plugins_1.HARDHAT_NETWORK_NAME) {
        return await runSuper(taskArgs);
    }
    const polkadotGlobal = global;
    const { commandArgs, server, port } = await (0, utils_1.startServer)({
        forking: env.config.networks.hardhat.forking,
        forkBlockNumber: env.config.networks.hardhat.forking?.blockNumber,
        nodeCommands: env.userConfig.networks?.hardhat?.nodeConfig,
        adapterCommands: env.userConfig.networks?.hardhat?.adapterConfig,
    });
    try {
        await server.listen(commandArgs.nodeCommands, commandArgs.adapterCommands, false);
        await (0, utils_1.waitForNodeToBeReady)(port);
        const oldNetwork = env.network;
        await (0, utils_1.configureNetwork)(env.config, env.network, port);
        env.injectToGlobal();
        polkadotGlobal._polkadotNodeNetwork = env.network;
        const result = await runSuper(taskArgs);
        env.network = oldNetwork;
        delete polkadotGlobal._polkadotNodeNetwork;
        env.injectToGlobal();
        return result;
    }
    finally {
        await server.stop();
    }
}
